### PR TITLE
Fix DASC storage-boundary validation

### PR DESCRIPTION
**Pull request by kaix-nv**
Addresses the fresh full-diff review findings on #2387.

Changes:
- derive calibration horizons, geometry, and masks from decay parameters canonicalized to the declared checkpoint storage dtype
- validate boundary masks with inverse storage/live-dtype intervals while preserving the strict `horizon > Wmax` rule
- fail closed when restoring DASC metadata onto a model with no supported GDN modules, while retaining stale-policy warnings for recoverable GDN geometry drift
- import `importlib.util` explicitly and correct the storage-tolerance documentation
- add FP16 boundary-flip and wrong-architecture restore regressions

Validation:
- `python -m pytest -q tests/unit/torch/sparsity/state_sparsity/test_dasc.py`: 24 passed, 1 skipped (optional Megatron dependency)
- state + weight + attention sparsity tests: 302 passed, 1 skipped
- pre-commit on all touched files: passed
- real Transformers `Qwen3NextGatedDeltaNet` CPU smoke: BF16-declared calibration, BF16 reload, and policy export round-trip passed

Commit is ED25519-signed and carries a matching Signed-off-by trailer.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved sparsity policy validation to account for rounding in both stored and live tensor data types.
  * Policies now tolerate valid decay-related rounding differences while rejecting masks that are inconsistent with current parameters.
  * Improved handling and messaging for unsupported model structures during state restoration.

* **Tests**
  * Added coverage for storage-precision threshold behavior and unsupported model structures.

### CODE REVIEW

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 1]
```diff
@@ -348,7 +364,10 @@ def build_dasc_policy(
 
     modules = _get_gdn_modules(model)
     horizons = _analyze_gdn_modules(
-        modules, epsilon=config.epsilon, static_gate_input=config.static_gate_input
+        modules,
+        epsilon=config.epsilon,
+        static_gate_input=config.static_gate_input,
+        storage_dtype=_STORAGE_DTYPES[config.decay_parameter_storage_dtype],
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:418]: Valid. Fixed in #2389. The public analyze_gdn_decay API now accepts the same optional decay_parameter_storage_dtype used by calibration, and the guide requires deriving evaluated masks plus retained/total counts with it. The FP16 boundary regression now exercises this public path.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 2]
```diff
@@ -93,7 +98,7 @@ def restore_dasc_model(model: nn.Module, config: DASCConfig, metadata: MetadataD
 
     try:
         validate_dasc_model_structure(model, policy)
-    except ApplyModeError as error:
+    except _DASCModelStructureMismatchError as error:
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py]: Valid. Fixed in #2389 by making the save boundary match restore: unsupported/missing GDN architecture errors now propagate on both paths, while only recoverable GDN geometry or decay drift remains serializable with a warning. A regression verifies modelopt_state fails after replacing the GDN layer with nn.Linear.

File: docs/source/guides/6_sparsity.rst [hunk 3]
```diff
@@ -183,7 +183,9 @@ policy remains serializable so it cannot block saving or composing other ModelOp
 :func:`~modelopt.torch.sparsity.state_sparsity.export_policy` rejects it until recalibration.
 Set ``decay_parameter_storage_dtype`` to the checkpoint dtype for ``A_log`` and ``dt_bias`` before
 calibration. Policy validation allows only the rounding introduced by that declared storage dtype
-and the live tensor dtype; the default ``float32`` keeps unconfigured policies strict.
+and the live tensor dtype, so the effective tolerance is the wider of the two. The default
+``float32`` adds no storage slack of its own; decay tensors that are live in BF16 or FP16 are still
+validated against that live dtype's rounding.
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply docs/source/guides/6_sparsity.rst]: Valid. Fixed in #2390. The guide now states that distinct storage/live inverse rounding bounds are composed sequentially and matching dtypes are deduplicated.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 4]
```diff
@@ -413,34 +432,42 @@ def validate_dasc_model_structure(model: nn.Module, policy: DASCPolicy) -> None:
...
-    for name, values in current_horizons.items():
+    for name, module in modules.items():
         layer = policy.layers[name]
-        retained = [head for head, horizon in enumerate(values) if horizon > policy.selected_wmax]
-        if retained != layer.retained_heads:
-            raise ApplyModeError(
-                f"DASC policy head mask does not match current decay parameters in layer {name!r}"
-            )
+        if not module.A_log.dtype.is_floating_point or not module.dt_bias.dtype.is_floating_point:
+            raise ApplyModeError("GDN A_log and dt_bias must use floating-point dtypes")
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Valid. Fixed in #2390 with one shared validator used by horizon computation and export. It checks non-empty 1D equal shapes, floating dtype, and finiteness before bound construction; a NaN export regression covers the false-comparison hazard.
